[WIP] Move sidebar into a shadow DOM

mohitsakhuja opened the pull request:
### Changes introduced

- [x] Replace the `iframe` for the sidebar and use ~~a shadow DOM instead~~ the actual `Sidebar` component itself

mohitsakhuja commented:
@oliversauter I have a few questions before proceeding with the implementation. 

- Do we _need_ to put the sidebar into a shadow DOM too, now that the ribbon is already into one? I mean.. that shadow DOM boundary on the ribbon already protects our CSS from bleeding into the host. Can we do without using a nested shadow DOM for the Sidebar and just removing the `iframe` and using the Sidebar as a React component itself?
- And do we want the mounting/unmounting behavior of the `iframe` to be applied to whichever implementation we go with (shadow DOM or React component)? 
- @digiOps From what I can tell, the `FrameCommunication` that has been set up is only there because the `Sidebar` component is not directly being rendered by the `Ribbon`'s React component. There's an `iframe` in between which is responsible for rendering the sidebar and hence the need to setup communication between them. However, if we render it directly through our `Ribbon` React component, we should be able to achieve the same functionality (which is being achieved right now through `FrameCommunication`) through just React and/or by using Redux in conjunction, right? Of course, any other methods would still be made remotely callable.
- Is the `Sidebar` component that is defined in `src/sidebar-overlay` being used anywhere apart from the sidebar which is triggered from the ribbon? I tried but I couldn't find any other place where it was being used.

mohitsakhuja commented:
> Do you mean the unmounting on tab change? I am not sure, but an opened sidebar must stay opened so that a user does not lose their progress of writing something when they switch tabs.
> Is this afforded by keeping the current logic?

Yes and yes. In order to retain this, all that needs to be done is to use the actual `Sidebar` component instead of the `iframe`. The current logic has that feature implemented. 

> The problem is not only the bleeding, but also the performance. Iframes are pretty intensive, so we should definitely not have an iframe anywhere.

Of course. I'm all for ditching the `iframe`. It's just I'm wondering whether to replace it with a shadow DOM or to use the actual React component in its place.

> The sidebar component will also be used in the overview. Maybe some variation will be there between overview and in-page.
> Will do the mockups today so will know then.

Okay. Also, I'm not sure if it's already there, but even if it's not, we can still have a lower level component that both the in-page sidebar and the overview sidebar can extend from.

oliversauter commented:
> Of course. I'm all for ditching the iframe. It's just I'm wondering whether to replace it with a shadow DOM or to use the actual React component in its place.

To understand: the differentiation you make here is to look if the sidebar is a react component inside another shadow dom, or if it is a shadowdom in itself? I think for now a react component would be smoother to develop? We would not need to change so many things and also are not limited to just using html?

> Okay. Also, I'm not sure if it's already there, but even if it's not, we can still have a lower level component that both the in-page sidebar and the overview sidebar can extend from.

I think thats good.

mohitsakhuja commented:
> To understand: the differentiation you make here is to look if the sidebar is a react component inside another shadow dom, or if it is a shadowdom in itself? I think for now a react component would be smoother to develop? We would not need to change so many things and also are not limited to just using html?

Yes, the two options are:

1. Have the `Sidebar` React component inside a shadow DOM which is inside the `Ribbon` React component which is further inside a shadow DOM.
1. Have the `Sidebar` React component inside the `Ribbon` React component which is inside a shadow DOM.

And yes, going with the second option would be easier to develop. Also, it would allow everything to be inside the React tree (which of course is more manageable than having multiple RPC methods).

oliversauter commented:
> And yes, going with the second option would be easier to develop. Also, it would allow everything to be inside the React tree (which of course is more manageable than having multiple RPC methods).

What do you think are the performance differences of each variant?
It seems like we would not want to load the whole sidebar into each tab whenver we load the ribbon?

mohitsakhuja commented:
> What do you think are the performance differences of each variant?

I think the second option would provide _slightly_ better performance.

> It seems like we would not want to load the whole sidebar into each tab whenver we load the ribbon?

We won't be doing that. We'll retain the current logic being used for the `iframe` and use it for the `Sidebar` React component.

oliversauter commented:
> We won't be doing that. We'll retain the current logic being used for the iframe and use it for the Sidebar React component.

Ok great. 

I am currently chatting with a user who experiences quiet a bit of issues with the tabbing > assuming that too many processes are kicked off when a user changes the tabs. 
It's not a problem you need to solve right now, but to be aware of we don't want to load more processes in to that tab-changing. 

I will add a new issue to work on for the performance improvements.

mohitsakhuja commented:
> It's not a problem you need to solve right now, but to be aware of we don't want to load more processes in to that tab-changing.

Okay, I'll keep that in mind 🙂

mohitsakhuja commented:
This post is regarding the future implementation of the sidebar. 

After having seen a glimpse of what's coming in the future for the sidebar stuff - be it in-page or in the overview section, I have a bare-bones proposal for the implementation.

Currently, the code for the sidebar is structured like this:

- The sidebar lives inside an iframe which itself is a child of the ribbon.
- The way that sidebar and ribbon cooperate & communicate with each other is through the `FrameCommunication` class, which is a wrapper around the message passing available through the `window.postMessage` API.
- The sidebar takes in multiple kinds of props and depending on these props, it alters its behavior. 
- One of those props is the `env` variable which determines whether the sidebar is running `in-page` or in the `overview` section.

Now, after seeing the functionality that we want in the sidebar later down the road, I found similarities but there were quite a lot of differences as well. In my observation, the similarities involve individual components and not the whole sidebar itself. The differences in logic of the two lead me to believe that it might be better to:

- Split the logic of the sidebar for the in-page one and the one in the overview section. 
- Create smaller reusable components instead of reusing just one large component. 

This way, it would also allow us to make changes to the two sidebars separately, and give one of them unique features without cluttering the logic of the other.

@poltak @ShishKabab if you guys are okay with it, then I can go ahead and implement it. Since this will involve large changes, I think I'll first start with separating out the two sidebars (for the time being) and once having successfully converted the in-page sidebar to the new architecture, I'll begin with the overview one?

mohitsakhuja commented:
How should I proceed though? Should I copy the current directory for the sidebar to work on it? I tried doing it in place but I'm having a hard time changing the code in-place without breaking the `overview` sidebar..

oliversauter commented:
Thanks for laying out this so thoroughly. Much appreciated. 

I can't make any architectural decisions like this, but I am all in for sustainability. Especially under the perspective that we will add quite a lot of functionality/complexity to the sidebar in the time ahead. 

I will lay out all needed functionality as a checklist for you in [the Notion file](https://www.notion.so/worldbrain/Sidebar-Component-b8ddd3df512c4e318277224a9579c627).

poltak commented:
Not to belittle the work that has been put into it so far but, each time I had to look into the code, I also felt things were very complicated and hard to understand, especially with how the data and interscript communication flows happen (although there are lots of other complicated parts of our codebase too). IMO it would be great to take a step back, map it all out at a higher level, and see what could be done to simplify things.

I don't see any points I have an issue with.

> Regarding the highlights and annotations, we can either make the ribbon expose their functions but not control them, or we can maybe even move them to a content script of their own. I prefer the second option but I'm not yet sure how exactly to do it.

I'm not entirely sure on the main idea here (what exactly is being put into a content script?) but if you want to move things to another content script that's totally do-able but I think we'd have to alter the build slightly. I can assist with that if we go this route.

ShishKabab commented:
Hi @mohitsakhuja , thanks for all the research done so far! Having also interacted with the Ribbon code recently to implement a relatively simple feature, I also think it's time for a refactor before we move on to adding more features. Specifically it could be nice if you can:
* Implement your idea of splitting up big components to be used in different contexts. (Maybe you can use the `ui-logic` utils I've introduced here.)
* Investigate data flows. Data that needs to synchronized among multiple scripts (popup, sidebar, etc.) should live in the background script, and send signals to the right listening tabs if some change is requested.
* Seperate responsibilities and make good abstractions for communication. The sidebar should not be responsible for creating annotations, this should be in a separate content script module. Also, if there needs to be any inter-component communications, that should be done behind the screens. If a user requests the removal of a highlight from the sidebar, it should invoke a method on some object like `annatationManager.removeHighlight()` injected into React (props is fine, less magical / more explicit than context) which should know whether to send the request to the background script or somewhere else (AnnotationManager could be an abstract class implemented on way in the sidebar/popup, and another way in the background script, so there's a symmetry.) Also that manager should have a listenFor* method. I'd like to remove remoteFunction() calls directly from React, since it should be an implementation detail. This is just one example of creating clear boundaries of responsibilities, which can then be shuffled around more easily.
* Figure out if there's too many `if` statements between the different `env`s of the sidebar, and refactor where it makes sense into separate sidebars. You could make a `src/sidebar-common` dir containing re-usable components. One thing I've done with the `src/toolbar-notifications` was providing a default notification layout which every notification can, but is not required to use.

Hope this makes the work a bit more manageable  :)  The original code was written to ship quickly (thanks @digi0ps I remember you very well putting in a lot of work and being in close communication with us!), with the idea to refactor it later. Now I think it's a good moment to do that.

mohitsakhuja commented:
> I will lay out all needed functionality as a checklist for you in [the Notion file](https://www.notion.so/worldbrain/Sidebar-Component-b8ddd3df512c4e318277224a9579c627).

@oliversauter Thank you. That would be great!

> what exactly is being put into a content script?

@poltak I originally meant that we could put the highlighting/annotation functions that interact with the DOM into a content script. But now as @ShishKabab's has proposed - having an `AnnotationManager` class would be more elegant I guess. 

@ShishKabab thank you so much for writing out the different tasks, although I do feel a bit overwhelmed 😅

> The original code was written to ship quickly 

I have to give props to @digi0ps for being able to implement this, must have taken a lot of work! 👏

One last question though. Could you please suggest me what is the one thing that I should start from?

ShishKabab commented:
> One last question though. Could you please suggest me what is the one thing that I should start from?
I'd start by factoring out the communication between the components, getting them out of the React code. Then you can more easily fix to communication flow.

mohitsakhuja commented:
> I'd start by factoring out the communication between the components, getting them out of the React code. Then you can more easily fix to communication flow.

Alrighty! I'll start with that, thanks!

digi0ps commented:
Agreed that Ribbon's code is messy and complex. As @ShishKabab mentioned, the priority was on getting the features work back then, was planning for a refactor, but got stuck with academic stuff after that. 

Anyways:
- It would be great to take out the state from Ribbon and move out into Redux. Similarly the methods too following the points mentioned above. 
- Make submodules for Ribbon, Sidebar etc.
- You would be able to get rid of some `if` conditions based on `env` since iFrame put in more constraints ( Like some of the `menuStyles` were specific to the iFrame )

> We won't be doing that. We'll retain the current logic being used for the iframe and use it for the Sidebar React component.

I don't think we will need to implement the same logic used to stop rendering iFrames. AFAIK rendering the Sidebar component in React wouldn't produce as much load as rendering an iFrames. What do you think @poltak?

> I originally meant that we could put the highlighting/annotation functions that interact with the DOM into a content script

Most of the code which deals directly with the DOM live inside `interactions.js`.

>  Could you please suggest me what is the one thing that I should start from?

I would say moving the Ribbon state to Redux state would be a good starting point which doesn't break anything big. You can make a submodule for Ribbon with it's components and Redux stuff. There's already a store setup for in `src/sidebar-overlay/redux` which you can use.

On a side note, would love to help @mohitsakhuja if I have no assigned work after Onboarding, @oliversauter.

poltak commented:
> I don't think we will need to implement the same logic used to stop rendering iFrames. AFAIK rendering the Sidebar component in React wouldn't produce as much load as rendering an iFrames. What do you think @poltak?

I personally do not know what would produce more or less load. There shouldn't be any need to speculate though. Once we have the shadow DOM implementation working, without a great amount of effort, we should be able to compare them and make a decision based on what we observe.

oliversauter commented:
I found a weird behaviour in the current sidebar, make sure you keep a fix in mind when you refactor that part: 
When you do an annotation, it first tries to load all annotations for that page, and then shows the highlighted text. Instead it should display the highlight immediately (even if not saved yet in the backend so it is a smooth behaviour). http://recordit.co/eX67cIgW8g

mohitsakhuja commented:
@oliversauter that's actually a nice observation. Thanks for telling me about it. Will keep it in mind :+1:

oliversauter commented:
Another workflow improvement to keep in mind: 
- when a user is in edit mode of comments and clicks outside of the sidebar, the sidebar closes. (which is fine) however it also deletes all edits made without saving. So in this case it should always save the unsaved edits.

mohitsakhuja commented:
@oliversauter just wanted to ask you something - right now, when I go to edit an annotation through the sidebar and change the tags for it (add or delete, doesn't matter), but I click on 'Cancel' - the changes to the tags still get saved. Do we want this?

mohitsakhuja commented:
Upon further inspection I can see that we're setting the tags one by one for each annotation. When editing an annotation, it is certainly plausible that the user adds some new tags for it and deletes some old ones. The method being used above does take care of it, however as I've mentioned, it does come at the cost of not being able to cancel.

If we delay the process of actually saving the changes made by the user till the user has clicked 'Save', then that problem vanishes, however, the problem caused due to setting the tags one by one arises. When the user saves an edited annotation, we'll need to check for a few things:
1. Whether the edited annotation has at least one of the following two: an anchor and/or a comment with a non-zero length.
2. Find out which tags were deleted. Delete them from the storage one by one.
3. Find out which new tags were added. Add them to the storage one by one.

Apart from the first point, I think the other two are better handled by the annotation storage manager itself as it doesn't really make sense for the front-end to do that kind of checking.

I don't know how exactly the annotation storage has been implemented, but if it's possible then I think it would be better to have a `setAnnotationTags` RPC that would set the tags for a given annotation `url`. This would avoid the cumbersome one-by-one checking, I think. (Again, for the lack of my knowledge about how the annotation storage has been implemented, I'm just assuming it's possible to do so). Even if it does not avoid it, then at least it would move that checking away from the front-end. What do you say @digi0ps ?

oliversauter commented:
>  the changes to the tags still get saved. Do we want this?

I think ideally would be a real-time saving of things, but that would not go well with our append-only saving of changes for the backup. 

Re tags: why not checking that in the front end? seems to be easier on the system too? You already have the tag names, and if they changed then you would send a requet for either deletion or addition to the backend? how shall we do that @ShishKabab

poltak commented:
> I think it would be better to have a `setAnnotationTags` RPC that would set the tags for a given annotation url. This would avoid the cumbersome one-by-one checking

IMO this would be ideal. I don't see much reason to do multiple RPCs per tag added/deleted, apart from keeping the current behaviour of persisting that state when the user forgets to press "save". Having just one request with the state of the tags at the time of the "save" button press seems more ideal, as it minimizes all the RPCs to a constant and simplifies the "cancel" button's responsibilities.

Other possibility, to keep the current behaviour of 1 RPC per tag interaction and fix the "cancel" button - although more unnecessary DB writes and RPCs -, would be to push each op that has been called onto a stack to maintain order. Then "cancel" button just needs to pop everything off the stack and perform the inverse op.

For ref: the current annots storage implementation is in `modifyTags` method, which only accepts a single tag, hence why it sends a request each time the UI is interacted with:
https://github.com/WorldBrain/Memex/blob/develop/src/direct-linking/background/storage.ts#L269-L281

digi0ps commented:
Agreed, it would be more ideal to update the tags only when the user clicks the save button. 
The tag creation in the CommentBox follow a different behavior, where the tags are stored in a state array and after the annotation is created each tag is stored. 

One way I could think of is having an array for tags to add and tags to remove in the state. When the user clicks Save after editing, these array can passed to a remoteFunction which takes care of the operation.

mohitsakhuja commented:
See the thing with maintaining these additions and deletions in state is that I may add a tag and then remove it during the same edit session but that would make up for 2 unnecessary RPCs when I finally click 'Save'.

And I don't really think that the user would forget to press 'Save'. My reasoning for that is that the comments are supposed to be smaller in length, and hence the activity of saving the edit probably wouldn't escape from a person's mind during the edit period (in most cases). I doubt that most of the times they are going to be lengthy. If the edits are lengthy, then it makes sense to have incremental saves like word processors do. Because when you're writing a document, it's supposed to be a decent amount of length (a 600-800 character long document if it's about 200 words) and hence, it makes sense to have some sort of auto-save feature. 

With that said, I'll implement whichever solution that you guys deem appropriate. :)

oliversauter commented:
I notized a strange behaviour with the ribbon/sidebar. Wonder if that will be gone with both being a shadow dom implementation?
http://recordit.co/EfVnw8xEAg

Basically i can click on the ribbon elements even though they are behind the sidebar. Problem is here really that I can't "cancel" the comment being made. It then weirdly closes the sidebar or enables/Disables features of the ribbon.

oliversauter commented:
Some other workflow improvements to keep in mind: 
- [ ] When you edit the annotation note, then you cant use cmd+enter to save the note. Also the edit field does not have the same style as the initial comment field. (which it should have though)

mohitsakhuja commented:
> When you edit the annotation note, then you cant use cmd+enter to save the note. 

Just to be clear, what you're saying is we should be able to save the annotation using 'Cmd+Enter' when editing?

> Also the edit field does not have the same style as the initial comment field. (which it should have though)

And regarding the design of the edit field, I actually haven't changed the base design for it. Agreed that it's not pixel-perfect yet. 
But here's the design that is currently being used in production:
![image](https://user-images.githubusercontent.com/30660843/50891051-74f1f100-1421-11e9-86e1-0fff00adf52c.png)
And here's the design that's deployed on this branch:
![image](https://user-images.githubusercontent.com/30660843/50891135-a66abc80-1421-11e9-8ad3-fc19b07b15a8.png)

I'm not sure how to make it more similar on my own. If you could tell me how to, I'll do it :slightly_smiling_face: 

> Right now annotations are not sorted [...]

Yeah, it's on my radar to do so.

And thanks for clearing the rest of the stuff up too :slightly_smiling_face:

oliversauter commented:
> Just to be clear, what you're saying is we should be able to save the annotation using 'Cmd+Enter' when editing?

Yes

> And regarding the design of the edit field, I actually haven't changed the base design for it. Agreed that it's not pixel-perfect yet.

Forget about that for now, will change anyway because we wont have this here anymore: 
<img width="347" alt="screen shot 2019-01-09 at 12 49 59" src="https://user-images.githubusercontent.com/7870039/50897671-1de22100-140d-11e9-84b2-7419136a7660.png">
We would immediately create a card when people would like to add a comment, therefore the difference will not be as huge anymore.

oliversauter commented:
another bug to keep track of: 
- [ ] on FF when you first open a page and make an annotation, the annotation text is not loaded. Therefore when you save it only makes a comment, not an annotation.

mohitsakhuja commented:
> on FF when you first open a page and make an annotation, the annotation text is not loaded. Therefore when you save it only makes a comment, not an annotation.

Ohhh, weird. I'm unable to reproduce it. Could you show me how you're getting it?

oliversauter commented:
Gave it a test. Seems to work smooth except a few hickups: 
- [ ] when you make an annotation or a new comment, make the cursor jump into the text field right away. 
- [ ] The generated urls are not ideal. it uses these hashes to identify the annotations: "https://en.wikipedia.org/wiki/Application_software/#1547483759002". This means those urls can only be understood by Memex and produce new entries in the DB because of the new URL too. Ideally the implementation is just like before where it jumps to the page getting the info which annotation anchor it is looking for, without the hash. 
Also this address puts a '/' in front of the hash appendix, resulting in a wrong page being linked to where it does not find the annotation.
- [ ] when an annotation is created crate a new card automatically, like in the below screenshot. Its automatically saved as a highlight. Cursor is automatically put in the comment box. If user starts typing the comment box expands for a better view and the "save/cancel" buttons appear. If they press tab, then the tag selector is turned on. If they press 'esc', the sidebar closes. 
![image](https://user-images.githubusercontent.com/7870039/51132546-c4298e80-1832-11e9-9b88-8db193de86f2.png)

mohitsakhuja commented:
Thanks for the feedback! I'll start with the second point first.

mohitsakhuja commented:
I have some questions.

> when a user is in edit mode of comments and clicks outside of the sidebar, the sidebar closes. (which is fine) however it also deletes all edits made without saving. So in this case it should always save the unsaved edits.

- Do we really want this? I mean.. the user might actually be trying to cancel it but this would save the edit. Also, if it's the (accidental) deletion of unsaved edits that we want to avoid, then if I may, I would suggest - what we can do is.. when a click outside of the sidebar is registered, we can check (along with all other present conditions that we're checking) if any annotation is being edited atm. If yes, then we do not close the sidebar when user clicks outside it. If not, then we do. 

> on FF when you first open a page and make an annotation, the annotation text is not loaded. Therefore when you save it only makes a comment, not an annotation.

- Is this still occurring for you because I still haven't been able to reproduce this.

> change the functionality of the ribbon's "turn off/on highlighter" to "turn off/on highlights in text (and highlighter)" This way people can see their highlights in an article without opening the sidebar which can get in the way a lot of times.

- I don't fully understand this one. Currently that button is used to control whether or not the tooltip would be displayed on every page. But if we change the functionality to turning on/off the actual highlights being shown on the page as well as the highlighter itself, how would that work? I mean.. if we want to display actual highlights without opening the sidebar, using another button/icon in the ribbon would probably be a better idea as having highlights on/off doesn't (or at least it shouldn't) affect whether the tooltip is enabled or not. And if there would be too many buttons, we can get rid of the "Worldbrain logo" button? The arrow in the ribbon already opens the sidebar on its own, so no point in having another button do the same.

oliversauter commented:
> Do we really want this? I mean.. the user might actually be trying to cancel it but this would save the edit. Also, if it's the (accidental) deletion of unsaved edits that we want to avoid, then if I may, I would suggest - what we can do is.. when a click outside of the sidebar is registered, we can check (along with all other present conditions that we're checking) if any annotation is being edited atm. If yes, then we do not close the sidebar when user clicks outside it. If not, then we do.

That might also be a good solution, however saving would probably be an easier way no? 
If we go for the route of not letting the sidebar be closed then we would need to indicate somehow that they need to save it first. Maybe making the frame red and let the annotation shake a bit?

From a workflow pov I think Its smoother to just save it in case they close it?

mohitsakhuja commented:
> If a person is showing the highlights they probably also want to enable the highlighter.

That is true but the opposite may not be true. The user may want to have the highlighter enabled (in case they wanna make more annotations), but still want to have a clean screen i.e. no highlights being shown.

> From a workflow pov I think Its smoother to just save it in case they close it?

I suppose it is. I'll go ahead with it. If we ever consider it again, I personally would like that feature where the annotation frame goes red and shakes a bit (will do it on a branch that doesn't contain so many other changes (hopefully)).

oliversauter commented:
> I suppose it is. I'll go ahead with it. If we ever consider it again, I personally would like that feature where the annotation frame goes red and shakes a bit (will do it on a branch that doesn't contain so many other changes (hopefully)).

Ok, then lets do it with the shaker if you think that is better. Probably also prevents some heart attack moments :) I guess if people complain we can auto save it or make it a setting somewhere later. 

> That is true but the opposite may not be true. The user may want to have the highlighter enabled (in case they wanna make more annotations), but still want to have a clean screen i.e. no highlights being shown.

Mhh here I guess that will be the less cases. If they just want to highlight we will also have the context menu soon.

mohitsakhuja commented:
Ah, okay! I'll go ahead and do it. Just one last question: once the bug fixes are dealt with, do you think we should first merge what's already done and then continue the rest of the work (feature changes)? I feel like this already involves too many changes (Disclaimer: I do not know what the "usual" amount of changes is), and since we have other people starting with annotations search, and also Kaiz talked to me about starting with this branch, so that kind of makes it even more urgent to get this one first, right? What do you say?

mohitsakhuja commented:
Yes, of course! :slightly_smiling_face:

poltak commented:
There's so much changes here that it would take days to properly review everything. The diff is also big enough that I had to pause memex indexing as it was freezing up my browser trying to scroll around that page :P Instead I tried to do a quick look over most things. Here's some points based on things I noticed (i understand some of them are very broad):

1. I mostly like the `src/sidebar-common` feature dir and how its laid out, although what's the point of the `components/` dir? Could each of those component dirs be moved up a level, like `comment-box/`? Maybe a better way to pose the question is "Why is `comment-box/` separate to all the other components in `components/`?"
2. I feel the `sidebar-common/components/annotation-box/annotation-box-container` module is rather complicated because of all the component state reducers and checks. Do you think it's worth moving this component state to redux and moving all those methods and checks to redux reducers + selectors? 
3. Keep in mind semantics when choosing HTML elements to use. Rather than using `<div>` and `<span>`s by default, there's usually more appropriate elements available to use. One obvious example is `sidebar-common/components/close-button` - a clickable component - which uses a `<div>` rather than `<button>` or `<a>` (I understand there's lots of bad examples of this in our codebase already, but best not to add more)
4. The new onboarding breaks when you click on the "Go back to Dashboard" after you make your first annotation. That `browser.tabs.create` call in `CongratsMessage` doesn't seem to work. 
5. You will need to update this branch with the latest `develop` and resolve conflicts with any StorageManager-using code, updating them to storex calls (like those changes in `src/direct-linking/background/storage` module). The API is different, although not much. It may also make the rebase simpler if you squash down your commits a bit (43 might be a bit painful, but maybe not; give it a try). Feel free to send me a message, maybe in `#storage-manager`, if you want any help with this point.

mohitsakhuja commented:
>     "Why is `comment-box/` separate to all the other components in `components/`?"

Okay, yeah, that doesn't make sense. Some commits ago there were a couple other folders on the same level but as they got converted into components (previously containers), I moved them into components. I agree that the current structure doesn't make much sense. I'm thinking.. that I can move the `sidebar-container.ts` file along with the actions, reducer, selectors and its types into a separate folder that will also have the sidebar component (currently in `components`). And regarding the rest of the components currently in `components`, should I move all of them one level up, or should I keep the ones that are completely self-dependent (e.g. `congrats-message`, `empty-message`, `footer`, `loader` etc.) in the `components` directory itself and move the rest of them up? 

Also, regarding your next point - the reason why I didn't start with using Redux for the `annotation-box-container` component was that I thought that it didn't make much sense to have its place in the Redux store as half of the state that its trying to maintain (i.e. `commentText`, `tagsInput`) is only used during the edit mode. In any case, now I do see and realize that there are many other methods that have gotten involved, and I think it could be better to split this component into 3 components that are dependent on the `mode`, maybe even 2 as `delete` and `default` modes differ only in the footer. That way all the editing logic will be removed from the component (adding tag, deleting tag, changing comment text). And then, if it still seems that the `edit` mode component or this component is still somewhat complex, then I can definitely move its state to Redux.

> Keep in mind semantics when choosing HTML elements to use. [...] One obvious example is `sidebar-common/components/close-button` - a clickable component - which uses a `<div>` rather than `<button>` or `<a>`

Yeah, surely, I'll change that. And I'll also go through the rest of the `sidebar-common` and `sidebar-overlay` folders to change any other such occurrences.
 
>  The new onboarding breaks when you click on the "Go back to Dashboard" after you make your first annotation. That `browser.tabs.create` call in `CongratsMessage` doesn't seem to work.

Umm, I didn't change the internal code for this file. I only renamed it. But then again, I do have to pull some certain changes that have been introduced in this file. Thanks for pointing it out though. I'll keep this in mind when finishing things off.

>  You will need to update this branch with the latest `develop` and resolve conflicts with any StorageManager-using code, updating them to storex calls (like those changes in `src/direct-linking/background/storage` module). The API is different, although not much. It may also make the rebase simpler if you squash down your commits a bit (43 might be a bit painful, but maybe not; give it a try). Feel free to send me a message, maybe in `#storage-manager`, if you want any help with this point.

Okay, thanks. They are rebased to the point just before the Storage Manager was integrated, so I don't think it should be that much of a trouble now, but I'll see.

Lastly, thank you @poltak for your detailed feedback :slightly_smiling_face:

mohitsakhuja commented:
> we've just kept everything completely flat [...] 

Okay, I'll get them out of their own directories.

> it's pretty much infinitely scalable horizontally [...] Your call :)

Okay, yeah. Alright, I'll consider both the approaches once I split it into 2 (or 3) different components. So if we were to go with the route of shifting its state to the Redux store, how/what would it look like? I mean.. we would want to store the "edit" state for each annotation separately, and I can think of two ways. One would be to have the edit state along with the already present annotation state (mentioned in `src/sidebar-common/sidebar/types`, and whenever comment text is changed or tags are changed.. we change the edit state for that particular annotation. Or we can keep the edit state outside of the actual annotations. Or is there some other way? 

> It may be something to do with no longer having the separate sidebar.js content script output in the build script? Maybe that had the WebExt polyfill (the stuff on browser) included, while the standard content script output doesn't? TBH I don't know exactly, I'm only guessing as I've had trouble accessing WebExt APIs at browser in the content script before.

Oh! Yeah, I guess that could be it. Thanks for the heads up, I'll let you know if it still exists after rebasing and trying some stuff out on my own.

And thank you :slightly_smiling_face:

poltak commented:
> So if we were to go with the route of shifting its state to the Redux store, how/what would it look like? 

I would probably just do it by indexing each annot's state by their ID (so `url`). The actual state pointed to by each ID could be pretty much the same shape as what already exists in the component state. Then just set up a selector that takes an ID and returns the assoc. state. 
That index could be represented as an object with ID keys pointing to individual annots state values. If order is important, maintain an additional array of those IDs (great example of this, and just good reading in general, on this page: https://redux.js.org/recipes/structuring-reducers/normalizing-state-shape#designing-a-normalized-state).

mohitsakhuja commented:
### Tasks remaining in this PR:

- [x] When you edit the annotation note, then you cant use cmd+enter to save the note.
- [ ] Right now annotations are not sorted in any way. You don't have to implement it right now for the sake of focus (annotation search is more important) but keep in mind with your architecture that we can easily sort annotations by e.g. creation date or position on page.
- [x] when you make an annotation or a new comment, make the cursor jump into the text field right away.
- [x] When I click on an annotation card and it scrolls to the highlighted section, it should also give a bigger frame around the currently selected annotation card
- [x] Clicking on an highlighted annotation scrolls the sidebar to display the same annotation and vice-versa.
- [ ] Change some HTML elements and use the ones with better semantics.
- [ ] Get branch up-to-date with `develop`.



### Tasks to be done after this PR is merged:

- [ ] When a user is in edit mode of comments and clicks outside of the sidebar, the sidebar should not close. Instead, the annotation box's frame should go red and shake a bit to indicate the user to either save or cancel the edits.
- [ ] change the functionality of the ribbon's "turn off/on highlighter" to "turn off/on highlights in text (and highlighter)" This way people can see their highlights in an article without opening the sidebar which can get in the way a lot of times.
- [ ] When user clicks on highlighted text, it should open the sidebar with the respective annotation marked



### Tasks that need to be discussed further

> when an annotation is created crate a new card automatically, like in the below screenshot. Its automatically saved as a highlight. Cursor is automatically put in the comment box. If user starts typing the comment box expands for a better view and the "save/cancel" buttons appear. If they press tab, then the tag selector is turned on. If they press 'esc', the sidebar closes.

Will the comment box and this have the same style except the part that displays the highlighted annotation?

Also, this might require a bigger change in the architecture. Reason being - once we have saved the annotation, we would be re-fetching all the annotations from the storage, and we would need to make sure a few things:

1. This annotation appears at the top.
2. Comment box would need to "edit" this annotation this time, rather than "saving" it.
3. We do not render this annotation while mapping over the list of all annotations.

mohitsakhuja commented:
@poltak That was an amazing article, so firstly, thanks for that. Although I had already factored out the `annotation-box-container` according to its modes (default & delete in one, edit in the other) and I saw that its just the `edit` mode that requires that "edit" piece of state and the reducers for it. Once you look at it, it's actually pretty small. Sure, the rest of the component is also big but that's because the UI functionality of it is rather complicated (auto resizing textarea, tabbing back and forth, Cmd/Ctrl+Enter to save). So after considering that and the fact that having the `edit` state in the Redux store would require changing the store structure quite substantially, I've kept it in the annotations state for now.

oliversauter commented:
Checked some last things and see if they worked: 
- [ ] there is this weird delay after closing the sidebar where the ribbon takes time to load: http://recordit.co/mVeMDn5WFr
Not sure if that is caused by some issue, doesn't necessarily need to be 'Fixed'
- [ ] There is this "search" and "done" in the tag selector. the 'search' can be removed, the 'done' should be the checkmark we already use <img width="361" alt="screen shot 2019-01-30 at 12 46 05" src="https://user-images.githubusercontent.com/7870039/51979318-26b69780-248d-11e9-8c63-cf7060e277b1.png">
- [ ] The "feature not available yet" > "Learn more" button does not work; there should be a 'back' button in the bottom bar of the annotation card so people can return to their original annotation
- [ ] There is this errorlogs happening on every open of the sidebar. What do they mean?  <img width="839" alt="screen shot 2019-01-30 at 12 50 50" src="https://user-images.githubusercontent.com/7870039/51979606-d855c880-248d-11e9-940a-bd1f502cb825.png">
- [ ] Also some smaller Design stuff, but I can do it. 

Great work so far:)

@poltak I wonder if we can implement some garbage removed version for testing. I think not only the performance is suffering (Especially when doing things like indexing), but also I wonder if we get reliable RAM usages without that?

mohitsakhuja commented:
> there is this weird delay after closing the sidebar where the ribbon takes time to load

Alright, I've fixed this now.

> There is this "search" and "done" in the tag selector. 

I found out the cause for this and I guess I could use some guidance here. So here's the cause for it: These icons are rendered by a font that we're using.. `material-icons`. I'm not quite sure as to how and where this font is being injected from. But since it's injected somewhere outside of the shadow DOM, it's not being picked by it.

> The "feature not available yet" > "Learn more" button does not work

Fixed this too :) 

> there should be a 'back' button in the bottom bar of the annotation card so people can return to their original annotation.

The "Close Notification" button does the same though. Should I rename it?

> There is this errorlogs happening on every open of the sidebar. What do they mean? 

:fearful: Wow. But I'm unable to reproduce this.. Could you tell me the steps that you followed to get to this?

> On this page: https://app.piesync.com/signin it somehow bleeds the host CSS into the shadowdom? 

Investigating this now.

poltak commented:
> @poltak I wonder if we can implement some garbage removed version for testing. I think not only the performance is suffering (Especially when doing things like indexing), but also I wonder if we get reliable RAM usages without that?

Do you mean this implementation is much worse in perf and resource usage than our previous iframe implementation? I would have expected the opposite. Has there been a comparison and/or discussion about this somewhere to give me some more context? I cannot see anything related in this PR and I'm a bit confused about the context surrounding this question.

If that mechanism that currently exists in prod to stop rendering the iframe/shadow DOM when the tab is no longer active is still in this new implementation, I would expect that should take care of unnecessary use of resources (if that's what you meant by "garbage removed version"?).

mohitsakhuja commented:
> If that mechanism that currently exists in prod to stop rendering the iframe/shadow DOM when the tab is no longer active is still in this new implementation

I disabled that mechanism in this implementation, but it can be enabled easily if that's what's required

oliversauter commented:
> I cannot see anything related in this PR and I'm a bit confused about the context surrounding this question.

Sorry that was actually somewhat unrelated. I just run into this issue all the time when testing. When need to import stuff, and/or then when doing queries it does take forever to execute them because of that garbage problem. We talked about it a while back I think.  I just want to be able to test in a clean environment without needing to build the production version. 

> Do you mean this implementation is much worse in perf and resource usage than our previous iframe implementation? I would have expected the opposite. 

I think its slightly less hungry, but I have to make a direct comparison still. I just wanted to be sure that its not tainted by some garbage collection stuff compared to the production version.

oliversauter commented:
> but it can be enabled easily if that's what's required

Yeah sure, that should definitely happen. With the ribbon/sidebar we would still load react into memory right? If that is the case we might want to think about making the ribbon itself just pure HTML.

poltak commented:
@oliversauter Ah I think I know what you mean now. That's from the redux dev tools. To disable it without doing a production build, comment out these 3 lines:
https://github.com/WorldBrain/Memex/blob/develop/src/options/options.jsx#L13-L15
Then change L12 to : `const ReduxDevTools = undefined`

> Yeah sure, that should definitely happen. With the ribbon/sidebar we would still load react into memory right? If that is the case we might want to think about making the ribbon itself just pure HTML.

I feel like optimization stuff like this should only be done if there's an actual problem we've identified and we're trying to solve. We implemented the "stop rendering the iframe when the tab is no longer active" thing to solve a problem that definitely existed with the previous implementation but, AFAIK from this thread, we haven't identified any problems here **yet**. (not saying there's no problems that exist at all here - TBH I have no idea - just trying to stay on-track)

oliversauter commented:
works all well, just 2 oddities I noticed: 
- [ ] I can't reliably reproduce this on other pages: http://recordit.co/2zcmgHAmmm but it seems as if when I enter the sidebar it would re-order the annotations in there? Is there maybe a time sort that only applies when the sidebar is :onhover? 
- [ ] All the shadowDom stuff, (e.g. tooltip, ribbon, sidebar and onboarding notifications) dont work anymore on firefox. This is the error message: 
<img width="839" alt="screen shot 2019-02-15 at 19 28 19" src="https://user-images.githubusercontent.com/7870039/52887102-71632f80-316e-11e9-929a-06459c296e76.png">

mohitsakhuja commented:
@oliversauter Regarding the Firefox issue, I don't think this is occurring due to anything changed in the PR. I am absolutely positive that the tooltip and ribbon/sidebar worked on Firefox when I pushed changes the last time (i.e. a couple weeks ago when I last worked on this).

Were any changes introduced to the build process and merged into `develop` branch? For me, even that branch isn't working on Firefox.

oliversauter commented:
Ok I can reproduce the error on Firefox too. I think before we launch anything we need to go on bughunting on this one. :(

oliversauter commented:
And this one I'd like to bring in too.

oliversauter commented:
I will merge this now. Great work @mohitsakhuja 

So to your question recently about me testing the performance of the extension with and without shadowdom, with and without sidebar enabled or not. 

Its barely noticable to be honest and with each tab the extension loads between 50-100mb or ram, levelling off around 500-600mb. 

My suspicion is that there are some other memory leak going on that block so much memory, thus this needs to be investigated separatel. Will open up a new PR for that.
